### PR TITLE
Populate meta criteria in a generic way

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

- Base available meta types on *_types configurations;
- Refactor Search::addMetaLeftJoin() to make it generic as possible;
- Fix some bugs.

Main problem now is that testing all these meta criteria takes time.

Here is the count of available meta criteria with this change:
| Itemtype               | Before   | After     |
|:---------------------- |:-------- |:--------- |
| **Total**              | **1270** | **19933** |
| Appliance              | -        | 786       |
| Budget                 | -        | 594       |
| Cartridge              | -        | 54        |
| CartridgeItem          | -        | 95        |
| Certificate            | -        | 774       |
| Change                 | -        | 19        |
| ChangeTask             | -        | 19        |
| Cluster                | -        | 414       |
| Computer               | 320      | 818       |
| Consumable             | -        | 72        |
| ConsumableItem         | -        | 95        |
| Contact                | -        | 19        |
| Contract               | -        | 873       |
| DCRoom                 | -        | 210       |
| DeviceBattery          | -        | 49        |
| DeviceCase             | -        | 49        |
| DeviceControl          | -        | 49        |
| DeviceDrive            | -        | 49        |
| DeviceFirmware         | -        | 49        |
| DeviceGeneric          | -        | 49        |
| DeviceGraphicCard      | -        | 49        |
| DeviceHardDrive        | -        | 49        |
| DeviceMemory           | -        | 49        |
| DeviceMotherboard      | -        | 49        |
| DeviceNetworkCard      | -        | 49        |
| DevicePci              | -        | 49        |
| DevicePowerSupply      | -        | 49        |
| DeviceProcessor        | -        | 49        |
| DeviceSoundCard        | -        | 49        |
| Document               | -        | 1053      |
| Domain                 | -        | 668       |
| DomainRecord           | -        | 177       |
| Enclosure              | -        | 571       |
| Entity                 | -        | 19        |
| Group                  | -        | 454       |
| ITILFollowup           | -        | 19        |
| ITILSolution           | -        | 19        |
| Infocom                | -        | 553       |
| Item_DeviceBattery     | -        | 55        |
| Item_DeviceCase        | -        | 55        |
| Item_DeviceControl     | -        | 55        |
| Item_DeviceDrive       | -        | 55        |
| Item_DeviceFirmware    | -        | 55        |
| Item_DeviceGeneric     | -        | 55        |
| Item_DeviceGraphicCard | -        | 55        |
| Item_DeviceHardDrive   | -        | 55        |
| Item_DeviceMemory      | -        | 55        |
| Item_DeviceMotherboard | -        | 55        |
| Item_DeviceNetworkCard | -        | 55        |
| Item_DevicePci         | -        | 55        |
| Item_DevicePowerSupply | -        | 55        |
| Item_DeviceProcessor   | -        | 55        |
| Item_DeviceSensor      | -        | 55        |
| Item_DeviceSimcard     | -        | 350       |
| Item_DeviceSoundCard   | -        | 55        |
| KnowbaseItem           | -        | 19        |
| Line                   | -        | 309       |
| Location               | -        | 792       |
| Monitor                | 190      | 669       |
| Netpoint               | -        | 22        |
| NetworkEquipment       | -        | 644       |
| NetworkPort            | -        | 376       |
| OperatingSystem        | -        | 307       |
| PDU                    | -        | 307       |
| PassiveDCEquipment     | -        | 240       |
| Peripheral             | 190      | 711       |
| Phone                  | 190      | 648       |
| Printer                | 190      | 648       |
| Problem                | -        | 19        |
| ProblemTask            | -        | 19        |
| Project                | -        | 589       |
| ProjectTask            | -        | 19        |
| Rack                   | -        | 530       |
| Reminder               | -        | 19        |
| Reservation            | -        | 344       |
| Software               | 190      | 756       |
| SoftwareLicense        | -        | 437       |
| SoftwareVersion        | -        | 26        |
| State                  | -        | 569       |
| Supplier               | -        | 19        |
| Ticket                 | -        | 41        |
| TicketTask             | -        | 19        |
| User                   | -        | 495       |

### CODE REVIEW

File: tests/functionnal/Search.php [hunk 1]
```diff
@@ -524,29 +503,44 @@ public function testSearchAllMeta() {
...
 
-         // Search on all with multiple meta criteria
-         // Limit criteria count to 50 to prevent performances issues
-         // and also prevent exceeding of MySQL join limit.
-         foreach (array_chunk($metacriteria, 50) as $metacriteria_chunk) {
+         // Search with criteria related to multiple meta items.
+         // Limit criteria count to 5 to prevent performances issues (mainly on MariaDB).
+         // Test would take hours if done using too many criteria on each request.
+         // Thus, using 5 different meta items on a request seems already more than a normal usage.
+         foreach (array_chunk($first_criteria_by_metatype, 5) as $criteria_chunk) {
```
Comment (cedric-anne) [inline tests/functionnal/Search.php:537]: According to my tests, we could gain something like 30 seconds on MariaDB by reducing this limit to 3.

Comment (trasher) [reply tests/functionnal/Search.php:537]: Problem is you try to call one search query with all criteria, right? Indeed, this is not a use case.

I guess we could keep a maximum or 2 or 3 criteria; making sure each individual one is tested that way.

Comment (cedric-anne) [reply tests/functionnal/Search.php:537]: For now, all criteria are tested independently (one search query for each criteria), then, in a second time, first criteria of each metatype are tested by groups (one search query with 5 criteria).

Comment (trasher) [reply tests/functionnal/Search.php:537]: Ok, so I guess limiting here to 3 would be enough on tests runs

File: inc/search.class.php [hunk 2]
```diff
@@ -2146,59 +2145,70 @@ static function computeTitle($data) {
...
    }
 
 
+
    /**
     * @since 0.85
     *
     * @param $itemtype
+    *
+    * @deprecated 9.5.4
```
Comment (trasher) [inline inc/search.class.php:2206]:
```suggestion
    * @deprecated x.x.x
```

File: inc/search.class.php [hunk 3]
```diff
@@ -5174,244 +5191,238 @@ static function addLeftJoin($itemtype, $ref_table, array &$already_link_tables,
...
-                                  AND `glpi_items_softwareversions_$to_type`.`is_deleted` = 0)
-                           $LINK `$to_table`
-                              ON (`glpi_items_softwareversions_$to_type`.`items_id`
-                                       = `$to_table`.`id`
-                                  AND `glpi_items_softwareversions_$to_type`.`itemtype` = '$to_type' ".
-                                  getEntitiesRestrictRequest("AND", $to_table).") ";
-            }
-            break;
-      }
+      return $JOIN;
```
Comment (trasher) [inline inc/search.class.php:5425]:
```suggestion
```

**Review by trasher — APPROVED** [review]

File: tests/functionnal/Search.php (hunk 1 above)
Comment (cedric-anne) [reply tests/functionnal/Search.php:537]:
```suggestion
         foreach (array_chunk($first_criteria_by_metatype, 3) as $criteria_chunk) {
```

**Review by orthagh — APPROVED** [review]

**Pull request merged**